File: scripts/pr-diff.py
#!/usr/bin/env python3
import os
import sys
import subprocess
from PIL import Image, ImageChops, ImageDraw

base_sha = os.environ.get('BASE_SHA')
head_sha = os.environ.get('HEAD_SHA')
if not base_sha or not head_sha:
    print("Error: BASE_SHA and HEAD_SHA must be set.")
    sys.exit(1)

scale = 4

try:
    diff_output = subprocess.check_output(
        ['git', 'diff', '--name-only', base_sha, head_sha, '--', 'skins/**/*.png'],
        text=True
    ).strip()
except subprocess.CalledProcessError:
    sys.exit(0)

if not diff_output:
    sys.exit(0)

changed_files = diff_output.split('\n')
diff_images = []

for file_path in changed_files:
    if not os.path.exists(file_path):
        continue

    base_name = os.path.splitext(os.path.basename(file_path))[0]
    old_path = None

    try:
        subprocess.check_call(
            ['git', 'cat-file', '-e', f'{base_sha}:{file_path}'],
            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL
        )
        old_path = f'/tmp/old_{base_name}.png'
        with open(old_path, 'wb') as f:
            subprocess.check_call(
                ['git', 'show', f'{base_sha}:{file_path}'],
                stdout=f
            )
    except subprocess.CalledProcessError:
        pass

    output_filename = f'diff_{base_name}.png'
    final_path = f'/tmp/{output_filename}'

    try:
        if old_path:
            img_old = Image.open(old_path).convert('RGBA')
            img_new = Image.open(file_path).convert('RGBA')

            if img_old.size != img_new.size:
                img_old = img_old.resize(img_new.size, Image.Resampling.LANCZOS)

            new_size = (img_new.width * scale, img_new.height * scale)
            img_old = img_old.resize(new_size, Image.Resampling.NEAREST)
            img_new = img_new.resize(new_size, Image.Resampling.NEAREST)

            diff_img = ImageChops.difference(img_old, img_new)

            mask = Image.new('RGBA', img_new.size, (0, 0, 0, 0))
            draw = ImageDraw.Draw(mask)
            pixels = diff_img.load()
            for x in range(diff_img.width):
                for y in range(diff_img.height):
                    r, g, b, a = pixels[x, y]
                    if r > 10 or g > 10 or b > 10:
                        draw.ellipse(
                            (x-2, y-2, x+2, y+2),
                            fill=(255, 0, 0, 180)
                        )

            result = Image.alpha_composite(img_new, mask)
            result.save(final_path, 'PNG')

        else:
            img_new = Image.open(file_path).convert('RGBA')
            new_size = (img_new.width * scale, img_new.height * scale)
            img_new = img_new.resize(new_size, Image.Resampling.NEAREST)
            img_new.save(final_path, 'PNG')

        diff_images.append({
            'path': final_path,
            'filename': output_filename,
            'file_path': file_path
        })

    except Exception as e:
        print(f"Failed to process diff for {file_path}: {e}")

if not diff_images:
    sys.exit(0)

os.makedirs('diff-images', exist_ok=True)
for img in diff_images:
    os.rename(img['path'], f'diff-images/{img["filename"]}')

comment_body = "### Skin Visual Diff\n\n> Red areas indicate modified pixels.\n\n"
for img in diff_images:
    image_url = f"https://raw.githubusercontent.com/{os.environ.get('GITHUB_REPOSITORY')}/pr-artifacts/pr-{os.environ.get('PR_NUMBER')}/diff-images/{img['filename']}"
    comment_body += f"**{img['file_path']}**\n"
    comment_body += f"![{img['filename']}]({image_url})\n\n"

with open('pr_comment.md', 'w') as f:
    f.write(comment_body)

with open('artifact-files.txt', 'w') as f:
    for img in diff_images:
        f.write(f"diff-images/{img['filename']}\n")

print('Comment body and artifact file list written')

File: scripts/pr-diff.mjs
import { execSync } from 'child_process';
import fs from 'fs';
import path from 'path';

const baseSha = process.env.BASE_SHA;
const headSha = process.env.HEAD_SHA;
const scale = 4;

let diffOutput = '';
try {
    diffOutput = execSync(`git diff --name-only ${baseSha} ${headSha} -- 'skins/**/*.png'`).toString().trim();
} catch (e) {
    process.exit(0);
}

if (!diffOutput) {
    process.exit(0);
}

const changedFiles = diffOutput.split('\n');
let commentBody = '### Skin Visual Diff\n\n';
commentBody += '> Red areas indicate modified pixels.\n\n';

for (const file of changedFiles) {
    if (!fs.existsSync(file)) continue;

    const baseName = path.basename(file, path.extname(file));
    let oldPath = null;

    try {
        execSync(`git cat-file -e ${baseSha}:${file}`, { stdio: 'ignore' });
        oldPath = `/tmp/old_${baseName}.png`;
        execSync(`git show ${baseSha}:${file} > ${oldPath}`);
    } catch (e) {}

    const finalPath = `/tmp/final_${baseName}.png`;

    try {
        let command;
        if (oldPath) {
            const scaledOld = `/tmp/scaled_old_${baseName}.png`;
            const scaledNew = `/tmp/scaled_new_${baseName}.png`;
            const maskPath = `/tmp/mask_${baseName}.png`;

            command = [
                `convert "${oldPath}" -scale ${scale * 100}% "${scaledOld}"`,
                `convert "${file}" -scale ${scale * 100}% "${scaledNew}"`,
                // 将 -compose src 移到命令末尾
                `compare "${scaledOld}" "${scaledNew}" -highlight-color white -lowlight-color black -compose src "${maskPath}"`,
                `convert "${maskPath}" -fill "rgba(255,0,0,0.5)" -opaque white -transparent black "${maskPath}"`,
                `convert "${scaledNew}" "${maskPath}" -compose over -composite "${finalPath}"`
            ].join(' && ');
        } else {
            command = `convert "${file}" -scale ${scale * 100}% "${finalPath}"`;
        }
        execSync(command);

        const buffer = fs.readFileSync(finalPath);
        const base64 = buffer.toString('base64');

        commentBody += `**${file}**\n`;
        commentBody += `![${baseName}](data:image/png;base64,${base64})\n\n`;
    } catch (err) {
        console.error(`Failed to process diff for ${file}:`, err.message);
    }
}

fs.writeFileSync('pr_comment.md', commentBody);
console.log('Comment body written to pr_comment.md');